File: C++/Codejam/2013/qualif/tictac.cpp
/* 
 * Author: makhtar
 * https://code.google.com/codejam/contest/2929SZ86/dashboard#s=p0
 * Created on Februarv 28, 2016, 11:55 PM
 * g++ -g -std=c++11 tictac.cpp -o tictac
 */
#include <map>
#include <vector>
#include <queue>
#include <deque>
#include <algorithm>
#include <sstream>
#include <iostream>
#include <fstream>

#include <cmath>
#include <cstdlib>
#include <cstring>

using namespace std;
#define SZ 4

string
solve(vector<char> b[21]) {
    char vch[] = {'X', '0'};
    stringstream ss;
    for (auto ch : vch) {
        bool wind1 = true; // diagonal 1
        bool wind2 = true;

        for (int x = 0; x < SZ; x++) {
            bool gotRow = true; // horizontal = row
            bool gotCol = true; // vertical = col

            for (int y = 0; y < SZ; y++) {
                // Check rows & cols
                ss << b[x][y] << " " << b[y][x] ;
                 clog << ss.str();
               // clog << "ch= " << ch << endl;
                if ((b[x][y] != ch) && (b[x][y] != 'T')) gotRow = false;
                if ((b[y][x] != ch) && (b[y][x] != 'T')) gotCol = false;
            }
            clog << endl;
            ss.clear();
            if (gotRow || gotCol) {
                ss << ch << " won";
                return ss.str();
            }

            // Check Diagonals [0, 0] [1, 1]...
            if ((b[x][x] != ch) && (b[x][x] != 'T')) wind1 = false;

            if ((b[3 - x][x] != ch) && (b[3 - x][x] != 'T')) wind2 = false;
        }
        if (wind1 || wind2) {
            ss << ch << " won";
            return ss.str();
        }
    }

    for (int x = 0; x < SZ; x++) {
        for (int y = 0; y < SZ; y++)
            if (b[y][x] == '.') return "Game has not completed";
    }

    return "Draw";
}

int
main(int argc, char** argv) {
    int T = 0; // num of test cases
    cin >> T;
    clog << T << " test cases\n";

    int n = 21; //SZ * SZ;
    char ch;
  
    for (int c = 0; c < T; ++c) {
        vector<char> rows[n];
        int j = 0;
        for (int i = 0; i < n; i++) {
            ch = cin.get();
            if (!isspace(ch)) {
                rows[j].push_back(ch);
                j++;
            }
            cerr << ch ;
        }
        ch = cin.get();
        clog << rows[0].size() << endl;

        string resp = solve(rows);
        cout << "Case #" << c + 1 << ": " << resp << endl;     
    }

    return 0;
}

File: C++/codejam/2013/qualif/TicTac.py

# Codejam 2013  Qualification solution
# TicTac Toe https://code.google.com/codejam/contest/2270488/dashboard#s=a&a=0
# python TicTac.py < A-small-practice.in 

import sys

def solve(b):

  for c in ['X', 'O']:
    wind1 = True
    wind2 = True
    for x in range(4):
      winh = True
      winv = True
      for y in range(4):

        # cols, rows check
        if b[y][x]!=c and b[y][x]!='T': winv = False
        if b[x][y]!=c and b[x][y]!='T': winh = False
      if winh or winv: return c + ' won'

      # Diagnonal check
      if b[x][x]!=c and b[x][x]!='T': wind1 = False
      if b[3-x][x]!=c and b[3-x][x]!='T': wind2 = False
    if wind1 or wind2: return c + ' won'

  for x in range(4):
    for y in range(4):
      if b[y][x]=='.': return 'Game has not completed'

  return 'Draw'


numcases = int(sys.stdin.readline())

for casenum in range(1,numcases+1):
  board = []
  for i in range(0,5):
    board.append(sys.stdin.readline().strip())
  print 'Case #' + repr(casenum) + ': ' + solve(board)
